src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -11,184 +10,167 @@ namespace System.IO
-                                UnsafeNativeMethods.FILE_SHARE_READ |
-                                    UnsafeNativeMethods.FILE_SHARE_DELETE |
-                                    UnsafeNativeMethods.FILE_SHARE_WRITE,          // share mode
-                                null,                                              // security descriptor
-                                UnsafeNativeMethods.OPEN_EXISTING,                 // how to create
-                                UnsafeNativeMethods.FILE_FLAG_BACKUP_SEMANTICS |
-                                    UnsafeNativeMethods.FILE_FLAG_OVERLAPPED,      // file attributes
-                                new SafeFileHandle(IntPtr.Zero, false)             // file with attributes to copy
-                            );
-
-            if (IsHandleInvalid)
+            var defaultSecAttrs = default(Interop.mincore.SECURITY_ATTRIBUTES);
+            _directoryHandle = Interop.mincore.CreateFile(
+                lpFileName: _directory,
+                dwDesiredAccess: Interop.mincore.FileOperations.FILE_LIST_DIRECTORY,
+                dwShareMode: FileShare.Read | FileShare.Delete | FileShare.Write,
+                securityAttrs: ref defaultSecAttrs,
+                dwCreationDisposition: FileMode.Open,
+                dwFlagsAndAttributes: Interop.mincore.FileOperations.FILE_FLAG_BACKUP_SEMANTICS | Interop.mincore.FileOperations.FILE_FLAG_OVERLAPPED,
+                hTemplateFile: IntPtr.Zero);
+            if (IsHandleInvalid(_directoryHandle))
             {
+                _directoryHandle = null;
                 throw new FileNotFoundException(SR.Format(SR.FSW_IOError, _directory));
             }
 
-            _stopListening = false;
-            // Start ignoring all events that were initiated before this.
-            Interlocked.Increment(ref _currentSession);
-
-            // Attach handle to thread pool
+            // Create the state associated with the operation of monitoring the direction
+            AsyncReadState state;
+            try
+            {
+                // Start ignoring all events that were initiated before this, and 
+                // allocate the buffer to be pinned and used for the duration of the operation
+                int session = Interlocked.Increment(ref _currentSession);
+                byte[] buffer = AllocateBuffer();
+
+                // Store all state, including a preallocated overlapped, into the state object that'll be
+                // passed from iteration to iteration during the lifetime of the operation.  The buffer will be pinned
+                // from now until the end of the operation.
+                state = new AsyncReadState(session, buffer, _directoryHandle, ThreadPoolBoundHandle.BindHandle(_directoryHandle));
+                unsafe { state.PreAllocatedOverlapped = new PreAllocatedOverlapped(ReadDirectoryChangesCallback, state, buffer); }
```

Comment:
I'd thought about it but it didn't seem worth the code complication.  It's one delegate created per run of the FSW, and usually with an FSW you create one, start it, and that's it.